### PR TITLE
Use Manifest Labeller for helm.

### LINKED ISSUE #2074: Helm deployments do not add labels to PodTemplates
### Expected behavior
Skaffold should also label pods created via Deployment resources.

### Actual behavior
Skaffold adds labels to the `deployment.metadata.labels` but not to `deployment.spec.template.metadata.labels`.

### Information

- Skaffold version: HEAD

This clearly comes from this function where the live api-resource is patched: https://github.com/GoogleContainerTools/skaffold/blob/25f1b66402507440a12a0c1173806b190ff76eb8/pkg/skaffold/deploy/labels.go#L95
It's clear, that it only adds labels to the owning resource, but not to the pods it manages.

### Steps to reproduce the behavior

1. `cd examples/helm-deployment`
2. `skaffold dev`
3. `kubectl get deploy skaffold-helm -o jsonpath='{..labels}'` -> yields the expected set of skaffold labels:
    ```go
    map[app:skaffold-helm app.kubernetes.io/managed-by:skaffold-v0.28.0-80-g25f1b664 chart:skaffold-helm-0.1.0 cleanup:true docker-api-version:1.38 heritage:Tiller release:skaffold-helm skaffold-builder:local skaffold-deployer:helm skaffold-tag-policy:sha256 tail:true] map[app:skaffold-helm release:skaffold-helm]
    ```
4. `kubectl get pod skaffold-helm-XYZ -o jsonpath='{..labels}' yields _only the helm labels_
    ```go
    map[app:skaffold-helm pod-template-hash:6b75889977 release:skaffold-helm]
    ```

I think it would be preferrable, if also the dependent resources have the Skaffold labels.

### PR DESCRIPTION
**Pull request by tejal29**
fixes #2074 

Description of Changes:
1. Use pkg/skaffold/deploy/kubect  Manifest labeller to label deployed objects.
1. Use `helm get manifests` to get all the helm deployed objects and re-use manifest labeller.
1. Add kubectl apply command after setting labels
1. Integration tests for helm are back
   a. #1823: Use env template in skaffold release name
1. During helm deploy, build is assumed and hence if no build-artifacts
are supplied, it fails with following error
 "no build present for gcr.io/k8s-skaffold/skaffold-helm"
  Since Build and Deploy are now separate ( #922 for notes) use the
  image value as is if not present as build artifacts. 
1. **Adding Labels to Services is back**
     In #959, we removed adding labels to services for helm tool (only). The issue https://github.com/kubernetes/kubernetes/pull/68087 is now fixed. The current integration test i have added has a service and it works fine.

### CODE REVIEW

File: pkg/skaffold/deploy/util.go [hunk 1]
```diff
@@ -18,45 +18,30 @@ package deploy
...
-			logrus.Infof("error parsing object from string: %s", err.Error())
+			logrus.Infof("error reading line string: %s", err.Error())
 			continue
 		}
-		obj, err := parseRuntimeObject(namespace, doc)
-		if err != nil {
-			logrus.Infof(err.Error())
-		} else {
-			results = append(results, obj)
+		if i := bytes.Index(b, []byte("apiVersion")); i >= 0 {
```
Comment (corneliusweig) [inline pkg/skaffold/deploy/util.go]: This is a bug, I think. It currently works, because in all examples you put `apiVersion` at the very beginning. But this is also a perfectly valid k8s yaml resource:
```yaml
kind: Pod
metadata:
  name: pod
  labels:
    app: skaffold-helm
spec:
  container:
    image: some
apiVersion: v1
```
But the appended manifest would just be `apiVersion: v1`, instead of the full yaml.

File: pkg/skaffold/deploy/util_test.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+%s---`, serviceYaml)),
+			expected: [][]byte{
+				[]byte(serviceYaml),
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.description, func(t *testing.T) {
+			var expected kubectl.ManifestList
```
Comment (corneliusweig) [inline pkg/skaffold/deploy/util_test.go]: The ManifestList is not really needed. It can be simplified to
```patch
Index: pkg/skaffold/deploy/util_test.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- pkg/skaffold/deploy/util_test.go	(date 1557261755000)
+++ pkg/skaffold/deploy/util_test.go	(date 1557493736000)
@@ -22,7 +22,6 @@
 	"fmt"
 	"testing"
 
-	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/deploy/kubectl"
 	"github.com/GoogleContainerTools/skaffold/testutil"
 )
 
@@ -140,16 +139,12 @@
 
 	for _, test := range tests {
 		t.Run(test.description, func(t *testing.T) {
-			var expected kubectl.ManifestList
-			for _, b := range test.expected {
-				expected.Append(b)
-			}
 			actual := parseReleaseInfo(bufio.NewReader(bytes.NewBuffer(test.releaseBytes)))
-			if len(actual) != len(expected) {
-				t.Errorf("want %d artifacts, got %d", len(expected), len(actual))
+			if len(actual) != len(test.expected) {
+				t.Errorf("want %d artifacts, got %d", len(test.expected), len(actual))
 			} else {
 				for i := 0; i < len(actual); i++ {
-					testutil.CheckDeepEqual(t, string(expected[i]), string(actual[i]))
+					testutil.CheckDeepEqual(t, string(test.expected[i]), string(actual[i]))
 				}
 			}
 		})
```

File: pkg/skaffold/deploy/util_test.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	"bytes"
+	"fmt"
+	"testing"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/deploy/kubectl"
+	"github.com/GoogleContainerTools/skaffold/testutil"
+)
+
+var (
+	deploymentYaml = `apiVersion: extensions/v1beta1
```
Comment (corneliusweig) [inline pkg/skaffold/deploy/util_test.go]: The actual content of the yamls does not really matter for this test, right? In that case, I would prefer a very simplistic yaml instead of such a full-blown resource. WDYT?

File: pkg/skaffold/deploy/helm.go [hunk 4]
```diff
@@ -70,30 +71,38 @@ func (h *HelmDeployer) Labels() map[string]string {
...
-			return errors.Wrapf(err, "deploying %s", releaseName)
+			return errors.Wrapf(err, "deploying %s", name)
 		}
-
-		dRes = append(dRes, results...)
+		releases[i] = name
 	}
 
+	manifests := h.getManifestsFromReleases(ctx, releases)
+	manifests, err := manifests.SetLabels(merge(labellers...))
```
Comment (corneliusweig) [inline pkg/skaffold/deploy/helm.go]: This only patches the yaml manifest bytes. It does not yet apply the manifest. I also checked with the helm demo and no labels were applied.
I think you still need to do the live-patching which you removed below.

File: pkg/skaffold/deploy/helm_test.go [hunk 5]
```diff
@@ -634,3 +565,83 @@ func makeRunContext(helmDeploy *latest.HelmDeploy, force bool) *runcontext.RunCo
...
+
+var (
+	releaseMap = map[string]string{
+		"release-1": fmt.Sprintf(`---
+# Source: skaffold-helm/templates/service.yaml
+%s---
+# Source: skaffold-helm/templates/deployment.yaml
+%s---`, serviceYaml, deploymentYaml),
+		"release-2": fmt.Sprintf(`---
+		# Source: skaffold-helm/templates/service.yaml
```
Comment (corneliusweig) [inline pkg/skaffold/deploy/helm_test.go]: This got me quite confused during review
```suggestion
		# Source: skaffold-helm/templates/pod.yaml
```

**Comment by corneliusweig** [comment]
@tejal29  Given that kokoro did not fail, we probably need a test which checks the labeller for all deployers? WDYT?

File: pkg/skaffold/deploy/util.go (hunk 1 above)
Comment (tejal29) [reply pkg/skaffold/deploy/util.go]: This is actually parsing the output of `helm get manifest` we use the trust like we do in `kubectl get`
1.  "---" is the yaml separtor
2. we can rely in apiVersion being present at the top.

File: pkg/skaffold/deploy/util_test.go (hunk 2 above)
Comment (tejal29) [reply pkg/skaffold/deploy/util_test.go]: sure!

File: pkg/skaffold/deploy/util_test.go (hunk 3 above)
Comment (tejal29) [reply pkg/skaffold/deploy/util_test.go]: yes.  This is also used in helm_test.go and this is the output of `helm get manifest` for the example repo.

File: pkg/skaffold/deploy/helm.go (hunk 4 above)
Comment (tejal29) [reply pkg/skaffold/deploy/helm.go]: That could be true. i should call the patch command like we do for kubectl and kustomize.

File: pkg/skaffold/deploy/helm_test.go (hunk 5 above)
Comment (tejal29) [reply pkg/skaffold/deploy/helm_test.go]: will do

**Comment by tejal29** [comment]
> @tejal29 Given that kokoro did not fail, we probably need a test which checks the labeller for all deployers? WDYT?

Thank @corneliusweig. Yes i was going to add an integration test and also verify it manually for the examples. I will do it today or Monday :).

File: pkg/skaffold/deploy/util.go (hunk 1 above)
Comment (corneliusweig) [reply pkg/skaffold/deploy/util.go]: Alright, I'm convinced.

File: pkg/skaffold/deploy/helm.go (hunk 4 above)
Comment (corneliusweig) [reply pkg/skaffold/deploy/helm.go]: Yes, that should work. Good idea!

**Comment by tejal29** [comment]
@corneliusweig  and @balopat  this is ready for final review.

File: pkg/skaffold/deploy/util_test.go (hunk 2 above)
Comment (tejal29) [reply pkg/skaffold/deploy/util_test.go]: done

File: pkg/skaffold/deploy/util_test.go (hunk 3 above)
Comment (tejal29) [reply pkg/skaffold/deploy/util_test.go]: done! This was more for testing. I made the yaml slim

File: pkg/skaffold/deploy/helm_test.go (hunk 5 above)
Comment (tejal29) [reply pkg/skaffold/deploy/helm_test.go]: done

File: pkg/skaffold/deploy/helm_test.go [hunk 6]
```diff
@@ -301,11 +242,10 @@ func TestHelmDeploy(t *testing.T) {
 			builds:      testBuilds,
 		},
 		{
-			description: "deploy error unmatched parameter",
+			description: "deploy should not error for unmatched parameter",
```
Comment (corneliusweig) [inline pkg/skaffold/deploy/helm_test.go]: This behavioral change seems unrelated to labels. Why was that necessary?

File: integration/examples/helm-deployment/skaffold.yaml [hunk 7]
```diff
@@ -8,7 +8,7 @@ build:
 deploy:
   helm:
     releases:
-    - name: skaffold-helm
+    - name: skaffold-helm-{{.TEST_NS}}
```
Comment (corneliusweig) [inline integration/examples/helm-deployment/skaffold.yaml]: Aren't the examples synced to integration from time to time? In that case, the namespace suffix would be removed.

File: pkg/skaffold/deploy/helm.go [hunk 8]
```diff
@@ -399,22 +418,23 @@ func (h *HelmDeployer) deleteRelease(ctx context.Context, out io.Writer, r lates
...
 	}
-
 	paramToBuildResult := map[string]build.Artifact{}
 	for param, imageName := range params {
 		newImageName := util.SubstituteDefaultRepoIntoImage(h.defaultRepo, imageName)
-		build, ok := imageToBuildResult[newImageName]
+		b, ok := imageToBuildResult[newImageName]
 		if !ok {
-			return nil, fmt.Errorf("no build present for %s", imageName)
+			logrus.Warnf("no build present for %s", imageName)
```
Comment (corneliusweig) [inline pkg/skaffold/deploy/helm.go]: Can this be condensed into one warning?

File: pkg/skaffold/deploy/helm.go [hunk 9]
```diff
@@ -399,22 +418,23 @@ func (h *HelmDeployer) deleteRelease(ctx context.Context, out io.Writer, r lates
...
+		imageToBuildResult[b.ImageName] = b
 	}
-
 	paramToBuildResult := map[string]build.Artifact{}
 	for param, imageName := range params {
 		newImageName := util.SubstituteDefaultRepoIntoImage(h.defaultRepo, imageName)
-		build, ok := imageToBuildResult[newImageName]
+		b, ok := imageToBuildResult[newImageName]
 		if !ok {
-			return nil, fmt.Errorf("no build present for %s", imageName)
```
Comment (corneliusweig) [inline pkg/skaffold/deploy/helm.go]: This change of error handling seems unrelated to labels. Should this be moved to a different PR?

File: pkg/skaffold/deploy/helm.go [hunk 10]
```diff
@@ -70,30 +76,45 @@ func (h *HelmDeployer) Labels() map[string]string {
...
+	manifests := h.getManifestsFromReleases(ctx, releases)
+	manifests, err := manifests.SetLabels(merge(labellers...))
+	if err != nil {
+		event.DeployFailed(err)
+		return errors.Wrap(err, "setting labels in manifests")
+	}
 
-	labels := merge(labellers...)
-	labelDeployResults(labels, dRes)
+	err = h.kubectl.Apply(ctx, out, manifests)
```
Comment (corneliusweig) [inline pkg/skaffold/deploy/helm.go]: :+1:

File: integration/helm_test.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+		"skaffold.dev/deployer":        "helm",
+	}
+
+	// check if labels are set correctly for deploument
+	dep := client.GetDeployment(depName)
+	if d := extractLabels(expectedLabels, dep.ObjectMeta.Labels); d != "" {
+		t.Errorf("did not find expected labels for dep %s: %s", depName, d)
+	}
+
+	// check if atleast one pod has has correct labels set.
```
Comment (corneliusweig) [inline integration/helm_test.go]:
```suggestion
	// check if at least one pod has has correct labels set.
```

File: integration/helm_test.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+	// To fix #1823, we make use of env variable templating for release name
+	env := []string{fmt.Sprintf("TEST_NS=%s", ns.Name)}
+	depName := fmt.Sprintf("skaffold-helm-%s", ns.Name)
+	helmDir := "examples/helm-deployment"
+	skaffold.Deploy().InDir(helmDir).InNs(ns.Name).WithEnv(env).RunOrFailOutput(t)
+
+	client.WaitForDeploymentsToStabilize(depName)
+
+	expectedLabels := map[string]string{
+		"app.kubernetes.io/managed-by": True,
```
Comment (corneliusweig) [inline integration/helm_test.go]: Could you rename `True` to something more expressive? This got me quite confused.

File: integration/helm_test.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+		}
+	}
+	if !foundPodWithCorrectLabels {
+		t.Errorf("Could not find one pod with all the labels. See diff:\n%s", strings.Join(diffs, "\n"))
+	}
+
+	skaffold.Delete().InDir(helmDir).InNs(ns.Name).WithEnv(env).RunOrFail(t)
+}
+
+func extractLabels(expected map[string]string, actual map[string]string) string {
```
Comment (corneliusweig) [inline integration/helm_test.go]:
```suggestion
func diffLabels(expected map[string]string, actual map[string]string) string {
```

File: integration/helm_test.go [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+		if diffs[i] == "" {
+			foundPodWithCorrectLabels = true
+			break
+		}
+	}
+	if !foundPodWithCorrectLabels {
+		t.Errorf("Could not find one pod with all the labels. See diff:\n%s", strings.Join(diffs, "\n"))
+	}
+
+	skaffold.Delete().InDir(helmDir).InNs(ns.Name).WithEnv(env).RunOrFail(t)
```
Comment (corneliusweig) [inline integration/helm_test.go]: Should the cleanup be put into a `defer`ed function to avoid rogue deployments?

File: integration/examples/helm-deployment/skaffold.yaml (hunk 7 above)
Comment (tejal29) [reply integration/examples/helm-deployment/skaffold.yaml]: AFAIK, the script we have just makes sure for each directory in example, we have a directory in integration test.
I don't think we sync them from time to time. 
Having this in examples might break examples if users don't have it set.

File: integration/helm_test.go (hunk 12 above)
Comment (tejal29) [reply integration/helm_test.go]: sure, i can put a dummy version string.

File: integration/helm_test.go (hunk 14 above)
Comment (tejal29) [reply integration/helm_test.go]: will do

File: pkg/skaffold/deploy/helm_test.go (hunk 6 above)
Comment (tejal29) [reply pkg/skaffold/deploy/helm_test.go]: i added the context in the description #5 of the PR
```
5. During helm deploy, build is assumed and hence if no build-artifacts
are supplied, it fails with following error
"no build present for gcr.io/k8s-skaffold/skaffold-helm"
Since Build and Deploy are now separate ( #922 for notes) use the
image value as is if not present as build artifacts.
```

File: pkg/skaffold/deploy/helm.go (hunk 9 above)
Comment (tejal29) [reply pkg/skaffold/deploy/helm.go]: makes sense. opened a PR here. https://github.com/GoogleContainerTools/skaffold/pull/2121

**Comment by tejal29** [comment]
Thanks @corneliusweig 
Re:
>However, I am a bit worried about your findings that re-labeling the pods in the deployment's podTemplateSpec will redeploy the pods once more. Is this acceptable?

This is true. I also discovered this when testing. 


>Have you also considered patching the chart templates directly?
yes, thought about that, helm has a command `helm template --output-dir <>`
Then update all the manifests generated. 
But then i could not find a way to pass this to `helm install/update`

**Comment by corneliusweig** [comment]
@tejal29 Looking at the helm docs suggests that the `template` command is more for development and not for production (https://helm.sh/docs/helm/#helm-template):

> (...) any values that would normally be looked up or retrieved in-cluster will be faked locally.

Other options that come to my mind:
- Instead of patching the podTemplateSpec, we could directly patch the pods. Of course, this has the disadvantage of being transient, because the labels will be lost when a pod is evicted and recreated from the podTemplate. Nevertheless, this could be viable for `dev` mode.
- Wait for helm 3 :smile:  
- Patch the chart templates. This will be a mess though..

**Comment by tejal29** [comment]
Thanks @corneliusweig , so i digged into the helm code and helm uses KubeClient to deploy the generated manifest. Its all done via tiller service and probably wait untill helm 3 for tillerless helm!

**Comment by tejal29** [comment]
Re-open #2075  and get that in!

File: integration/util.go [hunk 15]
```diff
@@ -130,6 +130,15 @@ func (k *NSKubernetesClient) GetDeployment(depName string) *appsv1.Deployment {
 	return dep
 }
 
+// GetPods gets all pods in the client namespace
+func (k *NSKubernetesClient) GetPods() *v1.PodList {
+	pods, err := k.client.CoreV1().Pods(k.ns).List(meta_v1.ListOptions{})
+	if err != nil {
+		k.t.Fatalf("Could not find pods for in namespace %s", k.ns)
```
Comment (nkubala) [inline integration/util.go]: nit: "Could not find pods for ~in~ namespace"

File: pkg/skaffold/deploy/helm.go [hunk 16]
```diff
@@ -57,6 +59,10 @@ func NewHelmDeployer(runCtx *runcontext.RunContext) *HelmDeployer {
 	return &HelmDeployer{
 		HelmDeploy:  runCtx.Cfg.Deploy.HelmDeploy,
 		kubeContext: runCtx.KubeContext,
+		kubectl: kubectl.CLI{
```
Comment (nkubala) [inline pkg/skaffold/deploy/helm.go]: I'm a little confused by the alias you have above. you do 

```golang
kubectl    kubectl.CLI
```
but then here you have
```golang
kubectl: kubectl.CLI{
```
so is it actually `kubectl.CLI.CLI`? maybe we can just get rid of the alias?

File: pkg/skaffold/deploy/helm.go [hunk 17]
```diff
@@ -366,24 +388,25 @@ func (h *HelmDeployer) packageChart(ctx context.Context, r latest.HelmRelease) (
...
-	if err := h.helm(ctx, &releaseInfo, false, "get", release); err != nil {
+	if err := h.helm(ctx, &releaseInfo, false, "get", "manifest", release); err != nil {
 		return nil, fmt.Errorf("error retrieving helm deployment info: %s", releaseInfo.String())
 	}
 	return bufio.NewReader(&releaseInfo), nil
 }
 
 // Retrieve info about all releases using helm get
-// Skaffold labels will be applied to each deployed k8s object
+// Skaffold labels will be applied to each deployed k8s object including k8object from
```
Comment (nkubala) [inline pkg/skaffold/deploy/helm.go]: nit: `Skaffold labels will be applied to each deployed k8s object, including those created from remote charts`

**Comment by tejal29** [comment]
with #2568 this is not needed.
